File: client/src/combat/local-projectiles.ts
// M2 LOCAL-ONLY projectiles. Renders + simulates rockets/grenades client-side
// using the EXISTING shared ballistics (stepProjectile/computeExplosion) — no
// reimplementation. On detonation it resolves damage/knockback through hooks so
// the integration layer applies them to the dummy and the local player (rocket
// jump). The server takes this over in M3; this class disappears then.

import * as THREE from 'three';
import type { Vec3, ProjKind, TraceWorld, Projectile, ExplosionHit } from '@rivals/shared';
import {
  TUNING,
  makeProjectile,
  stepProjectile,
  computeExplosion,
  v3,
} from '@rivals/shared';
import type { CapsuleTarget } from './hitscan';
import type { Particles } from '../render/particles';

export interface ProjectileHooks {
  localPlayerId: number;
  targets(): CapsuleTarget[]; // explosion candidates this frame (dummy + local player)
  onDamage(id: number, amount: number): void; // apply damage to entity id
  onImpulse(id: number, impulse: Vec3): void; // apply knockback (localPlayerId -> rocket jump)
  onDetonate?(pos: Vec3): void; // detonation FX cue (e.g. spatial explosion audio)
}

// Pre-sized pool: more than enough in flight at once (rocket fires ~0.83/s,
// grenade 1 carried). Spawning past capacity overwrites the oldest live one.
const POOL = 16;

// Emit a smoke puff roughly every this many metres of rocket travel.
const SMOKE_SPACING = 0.6;

// Direct-hit slop: a rocket counts as a direct hit on a capsule when the
// detonation point sits within (radius + this) of the capsule core segment.
const DIRECT_HIT_SLOP = 0.25;

// ---- module scratch (zero-alloc hot path) ----
const _step = { detonated: false, point: v3() };
const _impulse = v3();

interface Slot {
  proj: Projectile;
  mesh: THREE.Mesh;
  active: boolean;
  smokeAccum: number; // metres travelled since last smoke puff
  lastX: number;
  lastY: number;
  lastZ: number;
}

export class LocalProjectiles {
  readonly object: THREE.Group;

  private readonly world: TraceWorld;
  private readonly particles: Particles;
  private readonly hooks: ProjectileHooks;

  private readonly slots: Slot[] = [];
  private nextId = 1;
  private head = 0; // ring cursor for overwrite-oldest on overflow

  // Reused explosion-candidate buffer: shared computeExplosion wants
  // PlayerCapsule[]; CapsuleTarget is structurally identical, so we copy the
  // per-frame target list into a stable array to avoid per-detonation alloc.
  private readonly players: CapsuleTarget[] = [];

  // Two materials, built once.
  private readonly rocketMat: THREE.MeshBasicMaterial;
  private readonly grenadeMat: THREE.MeshStandardMaterial;
  private readonly rocketGeo: THREE.CylinderGeometry;
  private readonly grenadeGeo: THREE.SphereGeometry;

  constructor(world: TraceWorld, particles: Particles, hooks: ProjectileHooks) {
    this.world = world;
    this.particles = particles;
    this.hooks = hooks;

    this.object = new THREE.Group();
    this.object.name = 'projectiles';

    this.rocketMat = new THREE.MeshBasicMaterial({ color: 0xffcf6a });
    this.grenadeMat = new THREE.MeshStandardMaterial({
      color: 0x3a4a30,
      roughness: 0.8,
      metalness: 0.1,
      flatShading: true,
    });
    // Rocket: a short tube laid along +Z (we orient it to velocity at spawn).
    this.rocketGeo = new THREE.CylinderGeometry(0.08, 0.08, 0.5, 8);
    this.rocketGeo.rotateX(Math.PI / 2); // align cylinder axis (+Y) to +Z
    this.grenadeGeo = new THREE.SphereGeometry(0.13, 8, 6);

    // Pre-allocate the pool: a dummy Projectile + a hidden mesh per slot.
    for (let i = 0; i < POOL; i++) {
      const proj = makeProjectile(0, 'rocket', v3(), v3(), 0, 0);
      proj.alive = false;
      const mesh = new THREE.Mesh(this.rocketGeo, this.rocketMat);
      mesh.visible = false;
      mesh.frustumCulled = false;
      this.object.add(mesh);
      this.slots.push({ proj, mesh, active: false, smokeAccum: 0, lastX: 0, lastY: 0, lastZ: 0 });
    }
  }

  // velocity = dir * TUNING[kind].projSpeed; grenades get a fuse.
  spawn(kind: ProjKind, origin: Vec3, dir: Vec3, ownerId: number): void {
    const slot = this.acquire();
    const speed = TUNING[kind].projSpeed;
    const fuse = kind === 'grenade' ? TUNING.grenade.fuse : 0;

    const p = slot.proj;
    p.id = this.nextId++;
    p.kind = kind;
    p.pos.x = origin.x;
    p.pos.y = origin.y;
    p.pos.z = origin.z;
    p.vel.x = dir.x * speed;
    p.vel.y = dir.y * speed;
    p.vel.z = dir.z * speed;
    p.ownerId = ownerId;
    p.fuse = fuse;
    p.alive = true;

    slot.active = true;
    slot.smokeAccum = 0;
    slot.lastX = origin.x;
    slot.lastY = origin.y;
    slot.lastZ = origin.z;

    // Swap the mesh geometry/material to match the kind.
    slot.mesh.geometry = kind === 'rocket' ? this.rocketGeo : this.grenadeGeo;
    slot.mesh.material = kind === 'rocket' ? this.rocketMat : this.grenadeMat;
    slot.mesh.position.set(origin.x, origin.y, origin.z);
    if (kind === 'rocket') {
      // Point the tube down its travel direction.
      slot.mesh.lookAt(origin.x + dir.x, origin.y + dir.y, origin.z + dir.z);
    } else {
      slot.mesh.rotation.set(0, 0, 0);
    }
    slot.mesh.visible = true;
  }

  update(dt: number): void {
    for (let i = 0; i < this.slots.length; i++) {
      const slot = this.slots[i];
      if (!slot.active) continue;
      const p = slot.proj;

      stepProjectile(p, this.world, dt, _step);

      // Move the mesh to the projectile's new position.
      slot.mesh.position.set(p.pos.x, p.pos.y, p.pos.z);

      // Smoke trail for rockets, every SMOKE_SPACING metres of travel.
      if (p.kind === 'rocket') {
        const dx = p.pos.x - slot.lastX;
        const dy = p.pos.y - slot.lastY;
        const dz = p.pos.z - slot.lastZ;
        slot.smokeAccum += Math.hypot(dx, dy, dz);
        slot.lastX = p.pos.x;
        slot.lastY = p.pos.y;
        slot.lastZ = p.pos.z;
        if (slot.smokeAccum >= SMOKE_SPACING) {
          slot.smokeAccum = 0;
          this.particles.smoke(p.pos);
        }
        // Keep the tube oriented to velocity.
        slot.mesh.lookAt(p.pos.x + p.vel.x, p.pos.y + p.vel.y, p.pos.z + p.vel.z);
      }

      if (_step.detonated) {
        this.detonate(slot);
      }
    }
  }

  private detonate(slot: Slot): void {
    const p = slot.proj;
    this.particles.explosion(p.pos);
    this.hooks.onDetonate?.(p.pos);

    // Build the candidate list from this frame's targets (stable array reuse).
    const targets = this.hooks.targets();
    this.players.length = 0;
    let directHitId: number | undefined;
    for (let i = 0; i < targets.length; i++) {
      const t = targets[i];
      this.players.push(t);
      // Direct hit (rocket only): detonation point within radius+slop of the
      // capsule's vertical core segment.
      if (p.kind === 'rocket' && directHitId === undefined) {
        if (this.withinCapsule(p.pos, t, DIRECT_HIT_SLOP)) directHitId = t.id;
      }
    }

    const hits: ExplosionHit[] = computeExplosion(
      p.kind,
      p.pos,
      p.ownerId,
      this.players,
      this.world,
      directHitId,
    );
    for (let i = 0; i < hits.length; i++) {
      const h = hits[i];
      if (h.damage > 0) this.hooks.onDamage(h.id, h.damage);
      _impulse.x = h.impulse.x;
      _impulse.y = h.impulse.y;
      _impulse.z = h.impulse.z;
      this.hooks.onImpulse(h.id, _impulse);
    }

    // Retire the slot.
    slot.active = false;
    slot.proj.alive = false;
    slot.mesh.visible = false;
  }

  // Retire every live projectile (e.g. on a session/match teardown) so stale
  // rockets/grenades never linger into the next match.
  clear(): void {
    for (const slot of this.slots) {
      slot.active = false;
      slot.proj.alive = false;
      slot.mesh.visible = false;
    }
  }

  // Distance from `point` to the capsule's vertical core segment < radius + slop.
  private withinCapsule(point: Vec3, t: CapsuleTarget, slop: number): boolean {
    const segBottom = t.center.y - t.halfHeight;
    const segTop = t.center.y + t.halfHeight;
    const cy = point.y < segBottom ? segBottom : point.y > segTop ? segTop : point.y;
    const dx = point.x - t.center.x;
    const dy = point.y - cy;
    const dz = point.z - t.center.z;
    const r = t.radius + slop;
    return dx * dx + dy * dy + dz * dz <= r * r;
  }

  // Grab a free slot, or overwrite the oldest live one (ring) if the pool is full.
  private acquire(): Slot {
    for (let i = 0; i < this.slots.length; i++) {
      if (!this.slots[i].active) return this.slots[i];
    }
    const slot = this.slots[this.head];
    this.head = (this.head + 1) % this.slots.length;
    return slot;
  }
}

File: godot/src/Client/Render/LocalProjectiles.cs
// Client-side rockets/grenades: renders + simulates using the EXISTING shared
// ballistics (StepProjectile/ComputeExplosion). On detonation it resolves
// damage/knockback through hooks. Ported from client/src/combat/local-projectiles.ts.
//
// OFFLINE: full local damage/knockback. ONLINE: a cosmetic predicted copy whose
// damage hooks no-op (server owns damage) but self-knockback (rocket-jump) stays.

using System;
using System.Collections.Generic;
using Godot;
using WebRivals.Shared;
using WebRivals.Shared.Sim;
using WebRivals.Client.Combat;
using static WebRivals.Client.GodotConv;

namespace WebRivals.Client.Render;

public interface IProjectileHooks
{
    int LocalPlayerId { get; }
    List<CapsuleTarget> Targets();
    void OnDamage(int id, double amount);
    void OnImpulse(int id, Vec3 impulse);
    void OnDetonate(Vec3 pos);
}

public sealed class LocalProjectiles
{
    public readonly Node3D Object;

    private const int Pool = 16;
    private const double SmokeSpacing = 0.6;
    private const double DirectHitSlop = 0.25;

    private sealed class Slot
    {
        public Projectile Proj;
        public MeshInstance3D Mesh;
        public bool Active;
        public double SmokeAccum;
        public double LastX, LastY, LastZ;
    }

    private readonly ITraceWorld _world;
    private readonly Particles _particles;
    private readonly IProjectileHooks _hooks;

    private readonly Slot[] _slots = new Slot[Pool];
    private int _nextId = 1;
    private int _head = 0;

    private readonly ProjectileStep _step = new() { Detonated = false, Point = VMath.V3() };
    private readonly List<PlayerCapsule> _players = new();

    private readonly Mesh _rocketMesh;
    private readonly Mesh _grenadeMesh;
    private readonly Material _rocketMat;
    private readonly Material _grenadeMat;

    public LocalProjectiles(ITraceWorld world, Particles particles, IProjectileHooks hooks)
    {
        _world = world;
        _particles = particles;
        _hooks = hooks;
        Object = new Node3D { Name = "Projectiles" };

        _rocketMat = new StandardMaterial3D { AlbedoColor = new Color(1f, 0.812f, 0.416f), ShadingMode = BaseMaterial3D.ShadingModeEnum.Unshaded };
        _grenadeMat = new StandardMaterial3D { AlbedoColor = new Color(0.227f, 0.290f, 0.188f), Roughness = 0.8f, Metallic = 0.1f };
        // Rocket: a short tube laid along +Z (height along Z via rotation).
        _rocketMesh = new CylinderMesh { TopRadius = 0.08f, BottomRadius = 0.08f, Height = 0.5f, RadialSegments = 8 };
        _grenadeMesh = new SphereMesh { Radius = 0.13f, Height = 0.26f, RadialSegments = 8, Rings = 6 };

        for (int i = 0; i < Pool; i++)
        {
            var proj = Projectiles.MakeProjectile(0, ProjKind.Rocket, VMath.V3(), VMath.V3(), 0, 0);
            proj.Alive = false;
            var mesh = new MeshInstance3D { Mesh = _rocketMesh, MaterialOverride = _rocketMat, Visible = false, CastShadow = GeometryInstance3D.ShadowCastingSetting.Off };
            Object.AddChild(mesh);
            _slots[i] = new Slot { Proj = proj, Mesh = mesh, Active = false };
        }
    }

    public void Spawn(string kind, Vec3 origin, Vec3 dir, int ownerId)
        => Spawn(kind == "rocket" ? ProjKind.Rocket : ProjKind.Grenade, origin, dir, ownerId);

    public void Spawn(ProjKind kind, Vec3 origin, Vec3 dir, int ownerId)
    {
        var slot = Acquire();
        double speed = kind == ProjKind.Rocket ? Tuning.Rocket.ProjSpeed : Tuning.Grenade.ProjSpeed;
        double fuse = kind == ProjKind.Grenade ? Tuning.Grenade.Fuse : 0;

        var p = slot.Proj;
        p.Id = _nextId++;
        p.Kind = kind;
        p.Pos.X = origin.X; p.Pos.Y = origin.Y; p.Pos.Z = origin.Z;
        p.Vel.X = dir.X * speed; p.Vel.Y = dir.Y * speed; p.Vel.Z = dir.Z * speed;
        p.OwnerId = ownerId;
        p.Fuse = fuse;
        p.Alive = true;

        slot.Active = true;
        slot.SmokeAccum = 0;
        slot.LastX = origin.X; slot.LastY = origin.Y; slot.LastZ = origin.Z;

        slot.Mesh.Mesh = kind == ProjKind.Rocket ? _rocketMesh : _grenadeMesh;
        slot.Mesh.MaterialOverride = kind == ProjKind.Rocket ? _rocketMat : _grenadeMat;
        slot.Mesh.Position = ToGd(origin);
        if (kind == ProjKind.Rocket)
            OrientToVelocity(slot.Mesh, origin, dir);
        else
            slot.Mesh.Rotation = Vector3.Zero;
        slot.Mesh.Visible = true;
    }

    public void Update(double dt)
    {
        for (int i = 0; i < _slots.Length; i++)
        {
            var slot = _slots[i];
            if (!slot.Active) continue;
            var p = slot.Proj;

            Projectiles.StepProjectile(p, _world, dt, _step);
            slot.Mesh.Position = ToGd(p.Pos);

            if (p.Kind == ProjKind.Rocket)
            {
                double dx = p.Pos.X - slot.LastX, dy = p.Pos.Y - slot.LastY, dz = p.Pos.Z - slot.LastZ;
                slot.SmokeAccum += Math.Sqrt(dx * dx + dy * dy + dz * dz);
                slot.LastX = p.Pos.X; slot.LastY = p.Pos.Y; slot.LastZ = p.Pos.Z;
                if (slot.SmokeAccum >= SmokeSpacing)
                {
                    slot.SmokeAccum = 0;
                    _particles.Smoke(p.Pos);
                }
                OrientToVelocity(slot.Mesh, p.Pos, p.Vel);
            }

            if (_step.Detonated) Detonate(slot);
        }
    }

    private void Detonate(Slot slot)
    {
        var p = slot.Proj;
        _particles.Explosion(p.Pos);
        _hooks.OnDetonate(p.Pos);

        var targets = _hooks.Targets();
        _players.Clear();
        int directHitId = -1;
        for (int i = 0; i < targets.Count; i++)
        {
            var t = targets[i];
            _players.Add(new PlayerCapsule { Id = t.Id, Center = t.Center, Radius = t.Radius, HalfHeight = t.HalfHeight });
            if (p.Kind == ProjKind.Rocket && directHitId == -1 && WithinCapsule(p.Pos, t, DirectHitSlop))
                directHitId = t.Id;
        }

        var hits = Projectiles.ComputeExplosion(p.Kind, p.Pos, p.OwnerId, _players, _world, directHitId);
        for (int i = 0; i < hits.Count; i++)
        {
            var h = hits[i];
            if (h.Damage > 0) _hooks.OnDamage(h.Id, h.Damage);
            _hooks.OnImpulse(h.Id, h.Impulse);
        }

        slot.Active = false;
        slot.Proj.Alive = false;
        slot.Mesh.Visible = false;
    }

    public void Clear()
    {
        foreach (var slot in _slots)
        {
            slot.Active = false;
            slot.Proj.Alive = false;
            slot.Mesh.Visible = false;
        }
    }

    private static bool WithinCapsule(Vec3 point, CapsuleTarget t, double slop)
    {
        double segBottom = t.Center.Y - t.HalfHeight;
        double segTop = t.Center.Y + t.HalfHeight;
        double cy = point.Y < segBottom ? segBottom : (point.Y > segTop ? segTop : point.Y);
        double dx = point.X - t.Center.X;
        double dy = point.Y - cy;
        double dz = point.Z - t.Center.Z;
        double r = t.Radius + slop;
        return dx * dx + dy * dy + dz * dz <= r * r;
    }

    private static void OrientToVelocity(MeshInstance3D mesh, Vec3 pos, Vec3 dir)
    {
        var d = ToGd(dir);
        if (d.LengthSquared() < 1e-6f) return;
        // CylinderMesh axis is +Y; align +Y to travel direction.
        var fwd = d.Normalized();
        var up = Mathf.Abs(fwd.Y) < 0.99f ? Vector3.Up : Vector3.Right;
        var right = up.Cross(fwd).Normalized();
        var newUp = fwd; // cylinder long axis
        var basis = new Basis(right, newUp, right.Cross(newUp).Normalized());
        mesh.GlobalTransform = new Transform3D(basis.Orthonormalized(), mesh.GlobalPosition);
    }

    private Slot Acquire()
    {
        for (int i = 0; i < _slots.Length; i++)
            if (!_slots[i].Active) return _slots[i];
        var slot = _slots[_head];
        _head = (_head + 1) % _slots.Length;
        return slot;
    }
}
